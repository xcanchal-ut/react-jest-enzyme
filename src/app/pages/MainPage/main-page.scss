#main-page {
    background-image: url('/statics/img/owl.jpg');
    background-position: center;
    background-size: cover;
    background-repeat: no-repeat;
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    .filter-layer {
        position: absolute;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        opacity: 0.25;
        transition: all .2s ease;
        &.red {
            background-color: #F44336;    
        }
        &.green {
            background-color: #4CAF50;    
        }
        &.blue {
            background-color: #3F51B5;    
        }
        &.yellow {
            background-color: #FFC107;    
        }
    }

    .filters-content {
        padding: 50px 50px 20px 0;

        h1, 
        ul, 
        .checkbox {
            font-family: Helvetica;
            box-sizing: border-box;
            text-align: right;
            color: white;
            z-index: 1;
            position: relative;
        }

         h1 {
            font-size: 60px;
            margin: 0 20px 10px 0;
        }
    }
}