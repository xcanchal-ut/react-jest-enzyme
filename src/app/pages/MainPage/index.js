import React, { Component } from 'react';
import { find } from 'lodash';
import Title from '../../components/Title';
import Filters from '../../components/Filters/';
import Checkbox from '../../components/Checkbox/';
import './main-page.scss';

export default class MainPage extends Component {
    constructor() {
        super();
        this.state = {
            filters: [
                { text: 'Red', className: 'red' },
                { text: 'Green', className: 'green' },
                { text: 'Blue', className: 'blue' },
                { text: 'Yellow', className: 'yellow' },
            ],
            selectedFilter: {},
            filtersEnabled: true,
        };
        this.selectFilter = this.selectFilter.bind(this);
        this.toggleFiltering = this.toggleFiltering.bind(this);
    }

    selectFilter(className) {
        if (this.state.filtersEnabled) {
            this.setState({ selectedFilter: find(this.state.filters, { className }) });
        }
        return false;
    }

    toggleFiltering() {
        this.setState({ filtersEnabled: !this.state.filtersEnabled });
    }

    render() {
        const { selectedFilter, filtersEnabled } = this.state;

        return (
            <div id="main-page">
                <div className={`filter-layer ${selectedFilter.className}`} />
                <div className="filters-content">
                    <Title title="Apply filter" />
                    <Filters 
                        filters={this.state.filters} 
                        selected={this.state.selectedFilter} 
                        enabled={this.state.enabled}
                        onClick={(className) => this.selectFilter(className)}
                    />
                    <Checkbox
                        defaultChecked={this.state.filtersEnabled}
                        label={`Filters ${this.state.filtersEnabled ? 'enabled' : 'disabled'}`}
                        onClick={this.toggleFiltering}
                    />
                </div>
            </div>
        );
    }
}
