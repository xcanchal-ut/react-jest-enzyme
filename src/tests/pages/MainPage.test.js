import React from 'react';
import { mount } from 'enzyme';
import MainPage from '../../app/pages/MainPage/';

// PAGE TEST
let wrapper;
beforeEach(() => {
    wrapper = mount(<MainPage />);   
});

// Rendering
it('Should render a Title component, with an h1 inside that contains "Apply filter" (snapshot approach)', () => {
    const Title = wrapper.find('Title');
    expect(Title).toMatchSnapshot(); // initial render
}); 

it('Should render the four filters defined in the page state (traditional approach)', () => {
    const FiltersComponent = wrapper.find('Filters');
    const filters = FiltersComponent.find('li');
    expect(FiltersComponent.prop('filters')).toEqual(wrapper.state().filters);
    expect(filters.length).toEqual(4);
});

// Interaction
it('Should apply the red filter when clicking on the "red" button (snapshot approach)', () => {
    const filterLayer = wrapper.find('.filter-layer');
    const redFilter = wrapper.find('Filters').find('li.red');
    expect(filterLayer).toMatchSnapshot(); // initial unfiltered state (no className)
    redFilter.simulate('click');
    expect(filterLayer).toMatchSnapshot(); // after selecting one filter (className should be 'red')
});

it('Should toggle the checkbox value when clicking it (snapshot approach)', () => {
    const CheckboxComponent = wrapper.find('Checkbox').find('input');
    expect(CheckboxComponent).toMatchSnapshot(); // initial enabled state
    CheckboxComponent.simulate('click');
    expect(CheckboxComponent).toMatchSnapshot(); // after disabling filters
});

it('Should toggle the checkbox value when clicking it (traditional approach)', () => {
     const CheckboxComponent = wrapper.find('Checkbox').find('input');
     expect(CheckboxComponent.prop('defaultChecked')).toBeTruthy(); // initial enabled state
     CheckboxComponent.simulate('click');
     expect(CheckboxComponent.prop('defaultChecked')).toBeFalsy(); // after disabling filters
}); 