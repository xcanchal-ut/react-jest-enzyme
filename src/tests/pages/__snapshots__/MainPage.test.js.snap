// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should apply the red filter when clicking on the "red" button (snapshot approach) 1`] = `
<div
  className="filter-layer undefined"
/>
`;

exports[`Should apply the red filter when clicking on the "red" button (snapshot approach) 2`] = `
<div
  className="filter-layer red"
/>
`;

exports[`Should render a Title component, with an h1 inside that contains "Apply filter" (snapshot approach) 1`] = `
<Title
  title="Apply filter"
>
  <h1>
    Apply filter
  </h1>
</Title>
`;

exports[`Should toggle the checkbox value when clicking it (snapshot approach) 1`] = `
<input
  defaultChecked={true}
  onClick={[Function]}
  type="checkbox"
/>
`;

exports[`Should toggle the checkbox value when clicking it (snapshot approach) 2`] = `
<input
  defaultChecked={false}
  onClick={[Function]}
  type="checkbox"
/>
`;
